src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
```

Comment:
I don't necessary want to think in term of what **I** want.
I only think that *if* we decide that we want to lazily initialize the **PipeInputStream** - and only have it disposed by the consumer - then there may not be much interest in have a **Pipe**.

Problem I see is that ```SshCommand.EndExecute(IAsyncResult asyncResult)``` and ```SshCommand.Execute()``` would instantiate the **PipeInputStream** (as they invoke ```SshCommand.Result``` which in turn would lazily instantiate the **PipeInputStream**), but do not directly expose the created **PipeInputStream** to the consumer.  Meaning, the consumer would not be aware that a **PipeInputStream** is created, and that he would need (well *need* is a big word) to dispose it (through ```SshCommand.OutputStream```).

A solution for this would be to modify both ```SshCommand.Result``` and ```SshCommand.Error``` to dispose the **PipeInputStream** if they are creating it.

```cs
public string Result
{
    get
    {
        if (_result == null)
        {
            _result = new StringBuilder();
        }

        if (_outputReadStream != null)
        {
            if (_outputReadStream.DataAvailable)
            {
                // do not dispose the StreamReader, as it would also dispose the stream
                // and we are not the owner of this stream
                var sr = new StreamReader(_outputReadStream, _encoding);
                _result.Append(sr.ReadToEnd());
            }
        }
        else if (_outputWriteStream != null && _outputWriteStream.DataAvailable)
        {
            using (var outputReadStream = new PipeInputStream(_outputWriteStream))
            using (var sr = new StreamReader(_outputReadStream, _encoding))
            {
                _result.Append(sr.ReadToEnd());
            }

        }

        return _result.ToString();
    }
}
```

Remarks/questions here:
- Do we want **Result** and **Error** to block until the command has finished execution?
  Currently, they do not block. What do users expect?
- If we do not want these to block, then we need a **DataAvailable** property that returns a value indicating whether there's data available on the stream to be read. (similar to ```NetworkStream.DataAvailable```)